pkg/runtime.v2/core/trainingruntime.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	jobsetv1alpha2 "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+	runtime "github.com/kubeflow/training-operator/pkg/runtime.v2"
+	fwkcore "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/core"
+	fwkplugins "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/plugins"
+	idxer "github.com/kubeflow/training-operator/pkg/runtime.v2/indexer"
+)
+
+var (
+	errorNotFoundSpecifiedTrainingRuntime = errors.New("not found TrainingRuntime specified in TrainJob")
+)
+
+type TrainingRuntime struct {
+	framework *fwkcore.Framework
```

Comment:
Should we call it plugins and not framework ?
E.g. every runtime has list of plugins that it can register: JobSet, PodGroupPolicy, TorchPolicy, etc.
Additionally, we have registry of plugins and runtimes.

Reply:
Actually, this is a framework since the framework is this whole of process (https://docs.google.com/presentation/d/1HyEsBa7hxWpIoBXaX6uECiB48FWB85SG1kx15mO8hug/edit#slide=id.g30596bfee76_0_0), and the framework has some extension points like ComponentBuilder and then we can add plugins like JobSet, PyTorch against the extension points.

Reply:
I see, make sense.